### PR TITLE
OpenCL support

### PR DESCRIPTION
**Pull request by aarontcopal2**

### CODE REVIEW

File: Makefile.in [hunk 1]
```diff
@@ -341,7 +341,9 @@ OPT_CUDA_LDFLAGS = @OPT_CUDA_LDFLAGS@
 OPT_CUPTI = @OPT_CUPTI@
 OPT_CUPTI_IFLAGS = @OPT_CUPTI_IFLAGS@
 OPT_CUPTI_LDFLAGS = @OPT_CUPTI_LDFLAGS@
+OPT_GOTCHA_LDFLAGS = @OPT_GOTCHA_LDFLAGS@
```
Comment (jmellorcrummey) [inline Makefile.in:344]: The OPT_GOTCHA... flags in the OpenCL branch should disappear. The master branch and develop branch already have flags for GOTCHA. We should not have two distinct sets of GOTCHA flags for the same purpose.

File: src/lib/prof-lean/hpcrun-gotcha.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+//******************************************************************************
+// macros
+//******************************************************************************
+
+
+#define GOTCHA_GET_TYPED_WRAPPEE(wrapper, type) (type) gotcha_get_wrappee(wrapper)
+#else
+#define GOTCHA_GET_TYPED_WRAPPEE(wrapper, type) (type) 0
+#endif
+#endif
```
Comment (jmellorcrummey) [inline src/lib/prof-lean/hpcrun-gotcha.h:64]: this needs a newline at the end of the file

**Comment by aarontcopal2** [comment]
I have made the changes @jmellorcrummey .

File: src/tool/hpcrun/sample-sources/opencl.c [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+//******************************************************************************
+// local includes
+//******************************************************************************
+
+#include "simple_oo.h"
+#include "sample_source_obj.h"
+#include "common.h"
+
+#include <hpcrun/control-knob.h>
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/sample-sources/opencl.c:82]: i doubt that all of these header files need to be included.

my strategy for pruning includes is to put an #if 0 ... #endif around all of the includes. try compiling and see what breaks. then add back a header file needed to resolve a reported compiler error. repeat until it compiles. delete the #if 0 ... #endif and all of the includes of headers that weren't necessary to successfully compile.

File: src/tool/hpcrun/sample-sources/opencl.c [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+// or otherwise) arising in any way out of the use of this software, even
+// if advised of the possibility of such damage.
+//
+// ******************************************************* EndRiceCopyright *
+
+//******************************************************************************
+// system includes
+//******************************************************************************
+
+#include <alloca.h>
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/sample-sources/opencl.c:48]: see the comment about pruning unused header files. do this for the system includes as well as the local includes.

File: src/tool/hpcrun/gpu/opencl/opencl-memory-manager.h [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  union {
+    cl_kernel_callback_t ker_cb;
+    cl_memory_callback_t mem_cb;
+  };
+} opencl_object_details_t;
+
+
+typedef struct opencl_object_t {
+  s_element_ptr_t next;
+  opencl_object_channel_t channel;
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/opencl/opencl-memory-manager.h:94]: this is wrong. you want channel to be a pointer to a channel.

File: src/tool/hpcrun/gpu/opencl/opencl-memory-manager.h [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+// type declarations
+//******************************************************************************
+
+typedef enum {
+  OPENCL_KERNEL_CALLBACK                     = 0,
+  OPENCL_MEMORY_CALLBACK                     = 1
+} opencl_object_kind_t;
+
+
+typedef struct opencl_object_channel_t {
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/opencl/opencl-memory-manager.h:79]: You should have only a forward declaration for this channel here

`typedef struct opencl_object_channel_t opencl_object_channel_t;`

the channel type should only be fully defined in the .c file.

File: src/tool/hpcrun/gpu/opencl/opencl-intercept.c [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+  if (!event) {
+    mem_info->isInternalClEvent = true;
+    eventp = &my_event;
+  } else {
+    eventp = event;
+    mem_info->isInternalClEvent = false;
+  }
+  clwritebuffer_t clEnqueueWriteBuffer_wrappee = GOTCHA_GET_TYPED_WRAPPEE(clEnqueueWriteBuffer_handle, clwritebuffer_t);
+  cl_int return_status = clEnqueueWriteBuffer_wrappee(command_queue, buffer, blocking_write, offset, cb, ptr, num_events_in_wait_list, event_wait_list, eventp);
+  ETMSG(CL, "registering callback for type: H2D. Correlation id: %"PRIu64 "", correlation_id);
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/opencl/opencl-intercept.c:255]: I would prefer that you change the "CL" message to "OPENCL"

File: src/tool/hpcrun/gpu/opencl/opencl-memory-manager.c [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+opencl_object_t*
+opencl_malloc
+(
+  void
+)
+{
+  opencl_object_channel_t* c = opencl_object_channel_get();
+  opencl_object_t* cl_obj = channel_item_alloc(c, opencl_object_t);
+  cl_obj->channel = *c;
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/opencl/opencl-memory-manager.c:111]: this needs to be a pointer not a copy of the channel. see my comments in the header file.

**Comment by aarontcopal2** [comment]
@jmellorcrummey , the pull request has been updated.

File: src/tool/hpcrun/gpu/opencl/opencl-memory-manager.c [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+void
+opencl_free
+(
+  opencl_object_t* o
+)
+{
+  memset(o, 0, sizeof(opencl_object_t));
+  opencl_object_channel_t* c = &(o->channel);
+  channel_item_free(c, o);
+}
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/opencl/opencl-memory-manager.c:135]: add missing newline

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+#define OPENCL_FN_NAME(f) DYN_FN_NAME(f)
+
+#define OPENCL_FN(fn, args)                 \
+  cl_int (*OPENCL_FN_NAME(fn)) args
+
+#define HPCRUN_OPENCL_CALL(fn, args)        \
+{                                           \
+  cl_int status = OPENCL_FN_NAME(fn) args;  \
+  if (status != CL_SUCCESS) {               \
+    opencl_error_report(status);            \
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/opencl/opencl-api.c:103]: This does nothing. It returns an error string that is not printed. should an error cause the program to fail with an error?

Comment (aarontcopal2) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:103]: will this be ok?
```
if (status != CL_SUCCESS) {                                                 \
    ETMSG(OPENCL, "opencl call failed: %s", opencl_error_report(status));     \
  }
```

Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:103]: this is OK.

File: src/tool/hpcrun/gpu/opencl/opencl-api.c [hunk 11]
```diff
@@ -0,0 +1,460 @@
...
+(
+  void
+)
+{
+  ETMSG(OPENCL, "pending operations: %lu", atomic_load(&opencl_pending_operations));
+  while (atomic_load(&opencl_pending_operations) != 0);
+}
+
+
+static const char*
```
Comment (jmellorcrummey) [inline src/tool/hpcrun/gpu/opencl/opencl-api.c:204]: I don't like the fact that all of the OpenCL error code values are hardcoded into this file. Below is what I suggest instead. I thought that I posted this comment yesterday, but I don't see it or any change that would merit deleting it.

#define FORALL_OPENCL_ERRORS(macro)
  macro(CL_SUCCESS") \
  macro(CL_DEVICE_NOT_FOUND") \
  macro(CL_DEVICE_NOT_AVAILABLE") \
  macro(CL_COMPILER_NOT_AVAILABLE") \
  macro(CL_MEM_OBJECT_ALLOCATION_FAILURE") \
  macro(CL_OUT_OF_RESOURCES") \
  macro(CL_OUT_OF_HOST_MEMORY") \
  macro(CL_PROFILING_INFO_NOT_AVAILABLE") \
  macro(CL_MEM_COPY_OVERLAP") \
  macro(CL_IMAGE_FORMAT_MISMATCH") \
  macro(CL_IMAGE_FORMAT_NOT_SUPPORTED") \
  macro(CL_BUILD_PROGRAM_FAILURE") \
  macro(CL_MAP_FAILURE") \
  macro(CL_MISALIGNED_SUB_BUFFER_OFFSET") \
  macro(CL_EXEC_STATUS_ERROR_FOR_EVENTS_IN_WAIT_LIST") \
  macro(CL_COMPILE_PROGRAM_FAILURE") \
  macro(CL_LINKER_NOT_AVAILABLE") \
  macro(CL_LINK_PROGRAM_FAILURE") \
  macro(CL_DEVICE_PARTITION_FAILED") \
  macro(CL_KERNEL_ARG_INFO_NOT_AVAILABLE") \
  macro(CL_INVALID_VALUE") \
  macro(CL_INVALID_DEVICE_TYPE") \
  macro(CL_INVALID_PLATFORM") \
  macro(CL_INVALID_DEVICE") \
  macro(CL_INVALID_CONTEXT") \
  macro(CL_INVALID_QUEUE_PROPERTIES") \
  macro(CL_INVALID_COMMAND_QUEUE") \
  macro(CL_INVALID_HOST_PTR") \
  macro(CL_INVALID_MEM_OBJECT") \
  macro(CL_INVALID_IMAGE_FORMAT_DESCRIPTOR") \
  macro(CL_INVALID_IMAGE_SIZE") \
  macro(CL_INVALID_SAMPLER") \
  macro(CL_INVALID_BINARY") \
  macro(CL_INVALID_BUILD_OPTIONS") \
  macro(CL_INVALID_PROGRAM") \
  macro(CL_INVALID_PROGRAM_EXECUTABLE") \
  macro(CL_INVALID_KERNEL_NAME") \
  macro(CL_INVALID_KERNEL_DEFINITION") \
  macro(CL_INVALID_KERNEL") \
  macro(CL_INVALID_ARG_INDEX") \
  macro(CL_INVALID_ARG_VALUE") \
  macro(CL_INVALID_ARG_SIZE") \
  macro(CL_INVALID_KERNEL_ARGS") \
  macro(CL_INVALID_WORK_DIMENSION") \
  macro(CL_INVALID_WORK_GROUP_SIZE") \
  macro(CL_INVALID_WORK_ITEM_SIZE") \
  macro(CL_INVALID_GLOBAL_OFFSET") \
  macro(CL_INVALID_EVENT_WAIT_LIST") \
  macro(CL_INVALID_EVENT") \
  macro(CL_INVALID_OPERATION") \
  macro(CL_INVALID_GL_OBJECT") \
  macro(CL_INVALID_BUFFER_SIZE") \
  macro(CL_INVALID_MIP_LEVEL") \
  macro(CL_INVALID_GLOBAL_WORK_SIZE") \
  macro(CL_INVALID_PROPERTY") \
  macro(CL_INVALID_IMAGE_DESCRIPTOR") \
  macro(CL_INVALID_COMPILER_OPTIONS") \
  macro(CL_INVALID_LINKER_OPTIONS") \
  macro(CL_INVALID_DEVICE_PARTITION_COUNT") \
  macro(CL_INVALID_GL_SHAREGROUP_REFERENCE_KHR") \
  macro(CL_PLATFORM_NOT_FOUND_KHR") \
  macro(CL_INVALID_D3D10_DEVICE_KHR") \
  macro(CL_INVALID_D3D10_RESOURCE_KHR") \
  macro(CL_D3D10_RESOURCE_ALREADY_ACQUIRED_KHR") \
  macro(CL_D3D10_RESOURCE_NOT_ACQUIRED_KHR") 


#define FORALL_OPENCL_CALLS(macro) \
  macro(memcpy_H2D) \
  macro(memcpy_D2H) \
  macro(kernel)

#define CODE_TO_STRING(e) case e: return #e;



static const char*
opencl_call_to_string
(
  opencl_call_t type
)
{
  switch (type) {
  FORALL_OPENCL_CALLS(CODE_TO_STRING)
  default: return "CL_unknown_call";
  }
}


const char *
opencl_error_report
(
  cl_int error_status
)
{ 
  switch(error_status) {
  FORALL_OPENCL_ERRORS(CODE_TO_STRING)
  default: return "Unknown OpenCL error";
  }
}

Comment (aarontcopal2) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:204]: @jmellorcrummey . Got it. And I didnt see this comment yesterday, nor did I delete it

Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:204]: I may have added it and forgot to press the "comment" button.

Comment (aarontcopal2) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:204]: How is this solution any better? We are just wrapping the error strings inside macros

Comment (jmellorcrummey) [reply src/tool/hpcrun/gpu/opencl/opencl-api.c:204]: Notice that the numeric values of the codes and their binding with the error names are not embedded in the solution that I proposed.

**Comment by aarontcopal2** [comment]
@jmellorcrummey changes done. I will delete the above comments once these changes are accepted

**Comment by jmellorcrummey** [comment]
I made a few changes
- reformatted to eliminate long lines
- added an include file that wraps CL/cl.h and specifies verision 120
- adjusted use of include files to make sure that all .c files include their .h file
- tried to eliminate redundant includes

I did not build the result. please build and make sure that nothing was broken by my editing pass.

**Comment by aarontcopal2** [comment]
I took a pull and ran some examples. everything seems to work

**Pull request merged**